Exiled.CustomRoles/API/Features/ActiveAbility.cs

```diff
@@ -0,0 +1,141 @@
+// -----------------------------------------------------------------------
+// <copyright file="ActiveAbility.cs" company="Exiled Team">
+// Copyright (c) Exiled Team. All rights reserved.
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.CustomRoles.API.Features
+{
+    using System;
+    using System.Collections.Generic;
+
+    using Exiled.API.Features;
+
+    using MEC;
+
+    using UnityEngine;
+
+    using YamlDotNet.Serialization;
+
+    /// <summary>
+    /// The base class for active (on-use) abilities.
+    /// </summary>
+    public abstract class ActiveAbility : CustomAbility
+    {
+        /// <summary>
+        /// Gets or sets how long the ability lasts.
+        /// </summary>
+        public abstract float Duration { get; set; }
+
+        /// <summary>
+        /// Gets or sets how long must go between ability uses.
+        /// </summary>
+        public abstract float Cooldown { get; set; }
+
+        /// <summary>
+        /// Gets or sets an action to override the behavior of <see cref="CanUseAbility"/>.
+        /// </summary>
+        [YamlIgnore]
+        public virtual Func<bool> CanUseOverride { get; set; }
+
+        /// <summary>
+        /// Gets the last time this ability was used.
+        /// </summary>
+        [YamlIgnore]
+        public Dictionary<Player, DateTime> LastUsed { get; } = new Dictionary<Player, DateTime>();
+
+        /// <summary>
+        /// Gets all players actively using this ability.
+        /// </summary>
+        [YamlIgnore]
+        public HashSet<Player> ActivePlayers { get; } = new HashSet<Player>();
+
+        /// <summary>
+        /// Uses the ability.
+        /// </summary>
+        /// <param name="player">The <see cref="Player"/> using the ability.</param>
+        public void UseAbility(Player player)
+        {
+            ActivePlayers.Add(player);
+            LastUsed[player] = DateTime.Now;
+            ShowMessage(player);
+            AbilityUsed(player);
+            Timing.CallDelayed(Duration, () => EndAbility(player));
+        }
+
+        /// <summary>
+        /// Ends the ability.
+        /// </summary>
+        /// <param name="player">The <see cref="Player"/> the ability is ended for.</param>
+        public void EndAbility(Player player)
```

Comment:
This may cause issues if the plugin ends the ability before it ends. Maybe add an `ActivePlayers.Contains` check somewhere around here.